content/advent-2015/goaUntanglingMicroservices.md

```diff
@@ -0,0 +1,333 @@
+in Go would even be possible or yield something that is usable (that’s
+an area where dynamic languages definitely shine). goa started as an
+experiment but after many iterations of various degrees of ugliness
+I’m finally quite happy with the end result. Credits go to [Gomega](https://onsi.github.io/gomega/)
+for showing how using anonymous functions can help produce a clean and
+terse DSL. The goa DSL abstractions are taken directly from Praxis with
+a few tweaks to make them more amenable to the static nature and
+philosophy of the Go language.
+
+Here is what the DSL looks like for defining a type that can be used
+in the definition of request payloads or response media types:
+```go
+var BottlePayload = Type("BottlePayload", func() {
+	Attribute("name", func() {
+		MinLength(2)
+	})
+	Attribute("vineyard", func() {
+		MinLength(2)
+	})
+	Attribute("vintage", Integer, func() {
+		Minimum(1900)
+		Maximum(2020)
+	})
+	Attribute("color", func() {
+		Enum("red", "white", "rose", "yellow", "sparkling")
+	})
+})
+```
+This type can then be used when defining resource actions:
+```go
+var _ = Resource("bottle", func() {
+	DefaultMedia(Bottle)
+	BasePath("bottles")
+	Parent("account")
+	Action("create", func() {
+		Routing(POST(""))
+		Description("Record new bottle")
+		Payload(BottlePayload, func() {
+			Required("name", "vineyard")
+		})
+		Response(Created)
+	})
+})
+```
+As you can see the DSL is fairly self-descriptive. This last example
+also shows an interesting property which is that types can be referred
+to in different contexts and each context can add specific validations.
+Here the `create` action specifies that the `name` and `vineyard`
+fields of the `BottlePayload` data structure are required when the type
+is used in the payload (request body) of the `create` action for example.
+
+Obviously the DSL contains many more keywords, the point here was just
+to give you a sense of what it looks like. Should you want to know more
+consult the `dsl` package [GoDoc](https://godoc.org/github.com/raphael/goa/design/dsl).
+
+## The Magic: Code Generation
+
+goa comes with the `goagen` tool that runs the DSL which ends up
+producing simple data structures that represent the API design. These
+data structures describe the resources that make up the API and for
+each resource the actions complete with a description of their
+parameters, payload and responses. Note that the term *resource* here is
+very loosely defined. It's just a convenient way to group API endpoints
+(called *actions* in the DSL) together. The actual semantic is
+irrelevant to goa - goa is not an opinionated framework by design.
+
+`goagen` uses these data structures to generate many different outputs.
+The generated code takes care of validating the incoming requests and
+coercing the types to the ones described in the design, user code then
+only has to worry about the business logic.
+
+Here is a code snippet to illustrate the above, this code implements
```

Comment:
Maybe mention that this implementation is up to the user (skeleton is generated - look how easy it is to fill it in and make a working app!)